test/Graphics/Image.cpp

```diff
@@ -0,0 +1,151 @@
+#include <SFML/Graphics/Image.hpp>
+#include "GraphicsUtil.hpp"
+#include <array>
+
+#include <doctest.h>
+
+TEST_CASE("sf::Image - [graphics]")
+{
+    SUBCASE("Construction")
+    {
+        SUBCASE("Default constructor")
+        {
+            const sf::Image image;
+            CHECK(image.getSize() == sf::Vector2u());
+            CHECK(image.getPixelsPtr() == nullptr);
+        }
+    }
+
+    SUBCASE("Create")
+    {
+        SUBCASE("create(Vector2, Color)")
+        {
+            sf::Image image1;
+            image1.create(sf::Vector2u(10, 10), sf::Color::Red);
+
+            CHECK(image1.getSize() == sf::Vector2u(10, 10));
+            CHECK(image1.getPixelsPtr() != nullptr);
+
+            for (sf::Uint32 i = 0; i < 10; ++i)
+            {
+                for (sf::Uint32 j = 0; j < 10; ++j)
+                {
+                    CHECK(image1.getPixel(sf::Vector2u(i, j)) == sf::Color::Red);
+                }
+            }
+
+            sf::Image image2;
+            image2.create(sf::Vector2u(10, 10));
+            CHECK(image2.getSize() == sf::Vector2u(10, 10));
+            CHECK(image2.getPixelsPtr() != nullptr);
+
+            for (sf::Uint32 i = 0; i < 10; ++i)
+            {
+                for (sf::Uint32 j = 0; j < 10; ++j)
+                {
+                    CHECK(image2.getPixel(sf::Vector2u(i, j)) == sf::Color(0, 0, 0));
+                }
+            }
+        }
+
+        SUBCASE("create(Vector2, Uint8*)")
+        {
+            // 10 x 10, with 4 colour channels array
+            std::array<sf::Uint8, 400> pixels;
```

Comment:
Would it make more sense if this was a `std::array<sf::Color, 100>`? I believe that will still yield the exactly same data layout but with clearer semantics. Maybe even use an array of arrays to further improve semantics by making it abundantly clear that we're talking about a 10 by 10 array of colors.

Reply:
We need the `sf::Uint8` array here because that's one of the arguments for the create function it's testing

Reply:
```cpp
// 10 x 10, with 4 colour channels array
std::array<sf::Color, 100> pixels;
std::fill(pixels.begin(), pixels.end(), sf::Color(255, 0, 0, 255));

sf::Image image;
image.create(sf::Vector2u{10, 10}, &pixels.front().r);
```

Reply:
😢 Dang that's too bad. I thought it was a safe operation. Well that explains why `vector<Color>` isn't used in the implementation either.

Reply:
https://godbolt.org/z/1rG7xE7rE

Here's a Godbolt example of this problem if anyone wants to play around with it. It seems to work and I can't get any tools to emit diagnostics on this code, but that alone isn't sufficient evidence to ship code like this.

Reply:
I built an example program with AddressSanitizer and UndefinedSanitizer (macOS Monterey), neither of which reported any errors. I added unit tests that do this indexing and [all CI jobs passed](https://github.com/ChrisThrasher/SFML/actions/runs/2456576883). I'm running out of ways to demonstrate a failure.

Reply:
Are we okay with this change going into the tests, or should it be something that is updated after some `static_assert` is put into the `sf::Color` code?

Reply:
> I'm running out of ways to demonstrate a failure.

I think it was never about demonstrating, I see this more as a theoretical exercise. As mentioned, gamedev libraries have done stuff like this for decades. It works in practice (mostly 😉).

Reply:
No don't bother. I want to think on this construct more before we start using it.

Reply:
> _Disclaimer: my answer assumes that we want to do things_ correctly _and not apply the Yolo approach which the gamedev industry has been following for decades 😬_
> 
> What I found:
> 
> * https://stackoverflow.com/a/39432189
> * https://stackoverflow.com/a/44338825
> 
> From the second answer, I read that it is _explicitly defined behavior_ to reinterpret a type as `char`, `unsigned char` or `std::byte` (interestingly, they don't mention `signed char`). `sf::Uint8` is `unsigned char`.
> 
> So this might actually work. We should however do a `static_assert(sizeof(sf::Color) == 4)` (and maybe `alignof`) to prevent an exotic compiler from doing anything weird. It's extremely unlikely that `sf::Color` is padded, but IIRC not forbidden. What _is_ forbidden in C++ is reordering struct fields.

Good observation. This happens to be valid only because `sf::Uint8` is a `char`, which is a bit subtle. It also helps that this has been tested on all our CI machines, so I guess I don't have a big problem with this approach if we indeed add a static assertion to verify the alignment/size guarantees.